_posts/2023-10-23-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,257 @@
+Unfortunately, disabling topic creation doesn't help with applications that use the Kafka Admin client to directly create required topics (i.e. Kafka Streams based applications).
+In this case, the application should be written to wait for topic existence, or use an init container to do the same.
+
+### Finalizers
+
+[Finalizers](https://kubernetes.io/docs/concepts/overview/working-with-objects/finalizers) are used by default to avoid missing topic deletion events when the UTO is not running.
+A common pitfall is that the namespace becomes stuck in a "terminating" state when you try to delete it without first removing all finalizers.
+If this happens, you can simply remove finalizers from all `KafkaTopic` resources at once with the following command.
+
+```sh
+kubectl get kt -o yaml | yq 'del(.items[].metadata.finalizers[])' | kubectl apply -f -
+```
+
+It is also possible to disable finalizers using the following configuration.
+
+```sh
+# from Kafka custom resource
+spec:
+  entityOperator:
+    userOperator: {}
+    topicOperator: {}
+    template:
+      topicOperatorContainer:
+        env:
+          - name: STRIMZI_USE_FINALIZERS
+            value: "false"
+
+# from standalone Deployment
+spec:
+  template:
+    spec:
+      containers:
+        - name: STRIMZI_USE_FINALIZERS
+          value: "false"
+```
+
+### Upgrading from BTO to UTO
+
+The UTO is available as alpha feature since Strimzi 0.36.0, so we need to enable a feature gate in order to use it.
+[Strimzi 0.38.0](https://github.com/strimzi/strimzi-kafka-operator/releases/tag/0.38.0) adds Prometheus metrics and pause reconciliation feature.
+
+**This is not production ready yet, but the plan is to move it to beta in Strimzi 0.39.0, where it will be enabled by default.**
+**The UTO might require different CPU and memory resources, so we recommend to monitor and adjust accordingly.**
+
+```sh
+$ kubectl set env deploy strimzi-cluster-operator STRIMZI_FEATURE_GATES="+UnidirectionalTopicOperator"
+deployment.apps/strimzi-cluster-operator updated
+```
+
+If you are using the Operator Lifecycle Manager (OLM), then you simply need to patch the `Subscription`.
+
+```sh
+$ kubectl -n <namespace> patch sub <subscription-name> --type merge -p '
+  spec:
+    config:
+      env:
+        - name: STRIMZI_FEATURE_GATES
+          value: "+UnidirectionalTopicOperator"'
+subscription.operators.coreos.com/<subscription-name> patched
+```
+
+At this point the Cluster Operator will restart, and then it will deploy the new Entity Operator containing the UTO.
+You can check the logs to confirm that the UTO is running fine.
+
+```sh
+$ kubectl logs $(kubectl get po -l strimzi.io/name=my-cluster-entity-operator -o name) -c topic-operator | "Returning from main"
+2023-10-23 08:32:25,60913 INFO  [main] TopicOperatorMain:148 - Returning from main()
+```
```

Comment:
Aren't there settings printed at startup that would be more useful to grep for?

Reply:
Unfortunately no. This is something we can improve.